docs/cvl/cvl2/changes.md

```diff
@@ -0,0 +1,721 @@
+`address`.  This will cause undefined behavior, and in some cases the Prover
+will not be able to detect the error.
+````
+
+Changes to integer types
+------------------------
+
+In CVL 1, the rules for casting between integer types were complex; CVL 2
+simplifies them.
+
+The general rule of thumb is that you should use `mathint` for all data that is
+returned from contract functions, and the appropriate `int` or `uint` type for
+all data that will be passed as input to contract functions.
+
+It is now impossible for CVL math expressions to cause overflow - all integer
+operations are exact.
+
+### Mathematical operations return `mathint`
+
+In CVL 2, the results of all arithmetic operators have type `mathint`,
+regardless of the input types.  Arithmetic operators include `+`,
+`*`, `-`, `/`, `^`, and `%`, but not bitwise operators like `<<`, `xor`, and `|`
+(changes to bitwise operators are described {ref}`below <cvl2-bitwise>`).
+
+The primary impact of this change is that you may need to declare more of your
+variables as `mathint` instead of `uint`.  If you are performing arithmetic
+operations and integers that you are passing to specs, you will need to be more
+explicit about the overflow behavior by using the {ref}`new casting operators
+<cvl2-casting>`.
+
+```{todo}
+If you do not change this, you will see the following error:
+```
+
+### Comparisons require identical types
+
+When comparing two integers using `==`, `<=`, `<`, `>`, or `>=`, CVL 2 will
+require both sides of the equation to have identical types, and {ref}`implicit
+casts <cvl2-casting>` will not be used.  Comparisons with number literals (e.g.
+`0` or `1`) are allowed for any integer type.
+
+If you do not have identical types, the best solution is to use the special
+`to_mathint` operator to convert both sides to `mathint`.  For example:
+
+```cvl
+assert to_mathint(balanceOf(user)) == initial + deposit;
+```
+
+Note that in this example, we do not need to cast the right hand side, since
+the result of `+` is always of type `mathint`.
+
+````{todo}
+When should you not simply cast to `mathint`?  We have one example: consider the
+following code:
+
+```cvl
+ghost uint256 sum;
+
+hook ... {
+    havoc sum assuming sum@new == sum@old + newBalance - oldBalance;
+}
+```
+
+Simply casting to `mathint` will turn overflows into vacuity.  It's not clear
+how to generalize from this example.
+
+In this particular example, the right solution is to declare `sum` to be a
+`mathint` instead of a `uint`.  Note that with the more "modern" update syntax,
+this isn't a problem:
+
+```cvl
+ghost uint256 sum;
+
+hook ... {
+    sum = sum + newBalance - oldBalance;
+}
+```
+will say that the right-hand side is a `mathint` which can't be assigned to a
+`uint`.
+
+We hope that specification writers will think carefully about these changes and let
+us know of any other situations where the right thing is not simply casting to
+`mathint`.
```

Comment:
```suggestion
Simply casting to `mathint` will turn overflows into vacuity.

In this particular example, the right solution is to declare `sum` to be a
`mathint` instead of a `uint`.  Note that with the more "modern" update syntax,
this isn't a problem:

```cvl
ghost uint256 sum;

hook ... {
    sum = sum + newBalance - oldBalance;
}
```
will say that the right-hand side is a `mathint` which can't be assigned to a
`uint`.

We hope that specification writers will think carefully about these changes and let
us know of any other situations where the right thing is not simply casting to
`mathint`; it is not currently clear exactly how to generalize this example.
```